### PR TITLE
booking endpoint

### PR DESCRIPTION
**Pull request by gerimujo496**

### CODE REVIEW

File: middleware/isManager.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+const useContext = require("./useContext");
+
+module.exports = async(req, res, next) => {
+  const user = await useContext(req, res);
+  req.user = user;
+
+  if (user.isManager) next();
+  return res.status(400).send(`Invalid request !`);
```
Comment (Akeron44) [inline middleware/isManager.js]: Maybe 403 status would be better?
 return res.status(403).send("Access denied: No permission.");

File: middleware/validateParamsIDwithToken.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  return async (req, res, next) => {
+    const token = req.header("x-auth-token");
+    if (!token)
+      return res.status(401).send(`Access denied. No token provided.`);
+
+    const decoded = jwt.verify(token, process.env.JWT_PRIVATE_KEY);
+
+    const idFromParams = req.params[paramsNameToCompare];
+    if (decoded._id != idFromParams)
+      return res.status(403).send(`You don't have premission !`);
```
Comment (Akeron44) [inline middleware/validateParamsIDwithToken.js]: You don't have permission* or Permission denied.

File: middleware/isClient.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+const useContext = require("./useContext");
+const validateToken = require("./validateToken");
+
+module.exports = async (req, res, next) => {
+  const user = await useContext(req, res);
+  req.user = user;
+
```
Comment (KostandinD) [inline middleware/isClient.js:6]: Although what you are trying to achieve here with the middleware is correct, I suggest that next time do not implement it, as the isManager `middleware` would suffice for the use cases of this project.

File: middleware/isManager.js (hunk 1 above)
Comment (KostandinD) [reply middleware/isManager.js]: You could also consider 401 status error, which is Unauthorised.

File: middleware/useContext.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+module.exports = async (req, res) => {
+  const token = validateToken(req, res);
+
+  const decoded = jwt.verify(token, process.env.JWT_PRIVATE_KEY);
+
+  const user = await User.findOne({ _id: decoded._id });
+  if (!user) return res.status(400).send(`Invalid request !`);
+
+  return decoded;
```
Comment (KostandinD) [inline middleware/useContext.js]: I would suggest to have context as a separate function in another directory in order to access the context value. Also, the naming could suffice by being just `context`. You would need to cal it, so there is no need to be an anonymous function.

File: middleware/isClient.js [hunk 5]
```diff
@@ -0,0 +1,9 @@
+const useContext = require("./useContext");
+const validateToken = require("./validateToken");
```
Comment (KostandinD) [inline middleware/isClient.js]: I suggest that you should remove imports that you do not use.

File: middleware/validateToken.js [hunk 6]
```diff
@@ -0,0 +1,8 @@
+const jwt = require("jsonwebtoken");
```
Comment (KostandinD) [inline middleware/validateToken.js]: To the best of my understanding `validateToken` is used as a helper function. Therefore, I would suggest you remove it from the middleware function.

File: routes/booking.js [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    arrivalDate: { $gte: arrivalDate },
+    departureDate: { $lt: departureDate },
+    roomId,
+  };
+  const existingBooking = await Booking.findOne(existingBookingBody);
+
+  const room = await Room.findOne({ _id: roomId });
+  if (!room) res.status(400).send(`The room does not exists !`);
+
+  if (existingBooking && existingBooking.isApproved)
```
Comment (KostandinD) [inline routes/booking.js]: I would suggest you keep these conditions separated from one another. Maybe firstly you could check if the room exists and return the error accordingly. And the other error which checks if a room is approved within the provided time frame.

File: routes/booking.js [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    _id: bookingId,
+  };
+  const existingBooking = await Booking.findOne(bookingBody);
+
+  if (!existingBooking)
+    return res
+      .status(400)
+      .send(`The booking with the id: ${bookingId} does not exists  !`);
+
+  const bookingDelete = await Booking.deleteOne(bookingBody);
```
Comment (KostandinD) [inline routes/booking.js]: Maybe the naming could be `deletedBooking`, I think it would make more sense.

File: routes/booking.js [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+  };
+
+  const currentClient = await Booking.find(
+    queryBody.currentClientBody
+  ).populate(queryBody.populateUserId);
+
+  res.status(200).send(currentClient.userId);
+});
+
+router.get("/getBookingHistory/:userId", isClient, async (req, res) => {
```
Comment (KostandinD) [inline routes/booking.js]: Please, keep one source of truth and access the `userId` from the context and do not pass it via `queryParams`.

File: routes/booking.js [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+const express = require("express");
+const { Booking } = require("../models/booking");
+const { Room } = require("../models/room");
+const isClient = require("../middleware/isClient");
+const isManager = require("../middleware/isManager");
+const router = express.Router();
+
+router.post("/requestToBook/", isClient, async (req, res) => {
+  const { arrivalDate, departureDate, roomId } = req.query;
+  const { user } = req;
```
Comment (KostandinD) [inline routes/booking.js:13]: Once again, to keep up consistency, I would suggest to you to keep one source of truth, and use context to get the user info.

**Review by KostandinD — CHANGES_REQUESTED** [review]
The changes you have made clearly show that you have improved, but there are still certain points which you should refactor and focus on.

File: helpers/context.js [hunk 11]
```diff
@@ -0,0 +1,14 @@
+require("express-async-errors");
+const jwt = require("jsonwebtoken");
+const { User } = require("../models/user");
+const validateToken = require("./validateToken");
+
+module.exports = async (req, res) => {
```
Comment (KostandinD) [inline helpers/context.js:6]: Please, bare in mind to not declare a function as anonymous, but provide a naming for it.

File: models/user.js [hunk 12]
```diff
@@ -1,5 +1,6 @@
 const mongoose = require("mongoose");
 const jwt = require("jsonwebtoken");
+const { required } = require("joi");
```
Comment (KostandinD) [inline models/user.js:3]: Why do you need here `required` from Joi.

**Review by KostandinD — CHANGES_REQUESTED** [review]

File: routes/booking.js [hunk 13]
```diff
@@ -0,0 +1,341 @@
...
+
+  const booking = await Booking.findOne(queryBody.bookingBody)
+    .populate(queryBody.populateUserId)
+    .populate(queryBody.populateRoomId);
+
+  if (!booking)
+    return res
+      .status(400)
+      .send(`The booking with the id: ${bookingId} does not exists  !`);
+  booking;
```
Comment (Akeron44) [inline routes/booking.js:88]: Status should be 404 and the bookingId should not be included in the response for security reasons.

File: routes/booking.js [hunk 14]
```diff
@@ -0,0 +1,341 @@
...
+    roomId,
+    isApproved: null,
+  };
+  const booking = await Booking.findOne(queryBody.bookingBody);
+
+  if (!booking)
+    return res
+      .status(400)
+      .send(`The booking with the id: ${bookingId} does not exists  !`);
+
```
Comment (Akeron44) [inline routes/booking.js:122]: Status should be 404 and the bookingId should not be included in the response for security reasons.

**Review by Akeron44 — APPROVED** [review]

**Review by danielaKokoneshi309 — APPROVED** [review]

**Pull request merged**